File: platypus-js-web-client/src/main/java/com/eas/widgets/ImageParagraph.java
package com.eas.widgets;

import com.eas.ui.HasImageParagraph;
import com.eas.ui.Widget;
import com.google.gwt.dom.client.Document;
import com.google.gwt.dom.client.Element;
import com.google.gwt.dom.client.Style;
import com.google.gwt.user.client.ui.HasHTML;
import com.google.gwt.user.client.ui.HasText;

public abstract class ImageParagraph extends Widget implements HasText, HasHTML, HasImageParagraph {

    // forms api
    protected String text;
    protected boolean html;
    protected int horizontalTextPosition = RIGHT;
    protected int verticalTextPosition = CENTER;
    protected int iconTextGap = 4;
    protected int horizontalAlignment = LEFT;
    protected int verticalAlignment = CENTER;
    protected String image;
    protected Element content;
    protected Element aligner;

    protected ImageParagraph(Element aContainer, String aTitle, boolean asHtml) {
        this(aContainer, aTitle, asHtml, null);
    }

    protected ImageParagraph(Element aContainer, String aTitle, boolean asHtml, String aImage) {
        super(aContainer);
        text = aTitle;
        html = asHtml;
        image = aImage;
        element.getStyle().position = 'relative';
        element.getStyle().display ='inline-block');
        element.getStyle().setPadding(0+ 'px');
        //
        content = Document.get().createPElement();
        content.getStyle().setMargin(0+ 'px');
        content.getStyle().position = 'relative';
        content.getStyle().display ='inline-block');

        aligner = Document.get().createDivElement();
        aligner.getStyle().display ='inline-block');
        aligner.getStyle().position = 'relative';
        aligner.getStyle().height =100 + '%');
        aligner.getStyle().setVerticalAlign(Style.VerticalAlign.MIDDLE);
        aligner.getStyle().setVisibility(Style.Visibility.HIDDEN);
        //
        element.insertFirst(content);
        // aligner must go after content because of Gecko's craziness.
        element.insertAfter(aligner, content);
    }

    @Override
    public String getImageResource() {
        return image;
    }

    private void format() {
        if (isAttached()) {
            organizeText();
            organizeImage();
            organizeHorizontalAlignment();
            organizeVerticalAlignment();
        }
    }

    protected void organizeHorizontalAlignment() {
        final Style containerStyle = element.getStyle();
        final Style contentStyle = content.getStyle();
        switch (horizontalAlignment) {
            case LEFT:
            case LEADING:
                contentStyle.setLeft(0+ 'px');
                contentStyle.clearRight();
                contentStyle.setTextAlign(Style.TextAlign.LEFT);
                containerStyle.setTextAlign(Style.TextAlign.LEFT);
                break;
            case RIGHT:
            case TRAILING:
                contentStyle.clearLeft();
                contentStyle.right =0+ 'px');
                contentStyle.setTextAlign(Style.TextAlign.RIGHT);
                containerStyle.setTextAlign(Style.TextAlign.RIGHT);
                break;
            case CENTER:
                // contentStyle.setLeft(0+ 'px');
                // contentStyle.right =0+ 'px');
                contentStyle.setTextAlign(Style.TextAlign.CENTER);
                containerStyle.setTextAlign(Style.TextAlign.CENTER);
                break;
        }
    }

    protected void organizeVerticalAlignment() {
        final Style contentStyle = content.getStyle();
        final Style alignerStyle = aligner.getStyle();
        switch (verticalAlignment) {
            case TOP:
                contentStyle.setVerticalAlign(Style.VerticalAlign.TOP);
                break;
            case BOTTOM: {
                contentStyle.setVerticalAlign(Style.VerticalAlign.BOTTOM);
                break;
            }
            case CENTER: {
                contentStyle.setVerticalAlign(Style.VerticalAlign.MIDDLE);
                break;
            }
        }
    }

    protected void organizeText() {
        if (html) {
            content.setInnerHTML(text != null ? text : "");
        } else {
            content.setInnerText(text != null ? text : "");
        }
    }

    protected void organizeImage() {
        Style es = content.getStyle();
        int textGap = text != null && !text.isEmpty() ? iconTextGap : 0;
        if (image != null) {
            if (horizontalTextPosition == LEFT || horizontalTextPosition == LEADING) {
                /*
                backgroundPosition = "right";
                es.setPaddingLeft(0+ 'px');
                es.setPaddingRight(textGap + image.getWidth()+ 'px');
                */
            } else if (horizontalTextPosition == RIGHT || horizontalTextPosition == TRAILING) {
                /*
                backgroundPosition = "left";
                es.setPaddingLeft(textGap + image.getWidth()+ 'px');
                es.setPaddingRight(0+ 'px');
                */
            } else {
                if (text == null || text.isEmpty()) {
                    /*
                    int imageWidth = image.getWidth();
                    es.setPaddingLeft(imageWidth / 2+ 'px');
                    es.setPaddingRight(imageWidth / 2+ 'px');
                    */
                }
                //backgroundPosition = "center";
            }
            //backgroundPosition += " ";
            if (verticalTextPosition == TOP || verticalTextPosition == LEADING) {
                /*
                backgroundPosition += "bottom";
                es.setPaddingTop(0+ 'px');
                es.setPaddingBottom(textGap + image.getHeight()+ 'px');
                */
            } else if (verticalTextPosition == BOTTOM || verticalTextPosition == TRAILING) {
                /*
                backgroundPosition += "top";
                es.setPaddingTop(textGap + image.getHeight()+ 'px');
                es.setPaddingBottom(0+ 'px');
                */
            } else {
                if (text == null || text.isEmpty()) {
                    /*
                    int imageHeight = image.getHeight();
                    es.setPaddingTop(imageHeight / 2+ 'px');
                    es.setPaddingBottom(imageHeight / 2+ 'px');
                    */
                }
                //backgroundPosition += "center";
            }
            //es.setProperty("background", "url(" + image + ")" + " no-repeat " + backgroundPosition);
        }
    }

    @Override
    public int getVerticalAlignment() {
        return verticalAlignment;
    }

    @Override
    public void setVerticalAlignment(int aValue) {
        if (verticalAlignment != aValue) {
            verticalAlignment = aValue;
            format();
        }
    }

    @Override
    public int getHorizontalAlignment() {
        return horizontalAlignment;
    }

    @Override
    public void setHorizontalAlignment(int aValue) {
        if (horizontalAlignment != aValue) {
            horizontalAlignment = aValue;
            format();
        }
    }

    @Override
    public String getText() {
        return !html ? text : null;
    }

    @Override
    public void setText(String aValue) {
        html = false;
        text = aValue;
        format();
    }

    @Override
    public String getHTML() {
        return html ? text : null;
    }

    @Override
    public void setHTML(String aValue) {
        html = true;
        text = aValue;
        format();
    }

    @Override
    public int getIconTextGap() {
        return iconTextGap;
    }

    @Override
    public void setIconTextGap(int aValue) {
        if (iconTextGap != aValue) {
            iconTextGap = aValue;
            format();
        }
    }

    @Override
    public int getHorizontalTextPosition() {
        return horizontalTextPosition;
    }

    @Override
    public void setHorizontalTextPosition(int aValue) {
        if (horizontalTextPosition != aValue) {
            horizontalTextPosition = aValue;
            format();
        }
    }

    @Override
    public int getVerticalTextPosition() {
        return verticalTextPosition;
    }

    @Override
    public void setVerticalTextPosition(int aValue) {
        if (verticalTextPosition != aValue) {
            verticalTextPosition = aValue;
            format();
        }
    }

    @Override
    public void setImageResource(String aValue) {
        image = aValue;
        format();
    }
}

File: platypus-js-web-client/src/app/forms/image-paragraph.js
define([
    '../ui',
    '../extend',
    './widget'], function (
        Ui,
        extend,
        Widget) {
    function ImageParagraph(aElement, text, image, iconTextGap) {
        if (arguments.length < 4)
            iconTextGap = 4;
        if (arguments.length < 3)
            icon = null;
        if (arguments.length < 2)
            text = '';

        Widget.call(this, aElement);

        var self = this;

        var horizontalTextPosition = Ui.HorizontalPosition.RIGHT;
        var verticalTextPosition = Ui.VerticalPosition.CENTER;

        var paragraph = document.createElement('p');
        paragraph.classList.add('p-paragraph');
        this.element.appendChild(paragraph);

        this.element.classList.add('p-image-paragraph');

        function applyPosition() {
            self.element.classList.remove('p-image-paragraph-column');
            self.element.classList.remove('p-image-paragraph-row');
            self.element.classList.remove('p-image-paragraph-row-top');
            self.element.classList.remove('p-image-paragraph-row-bottom');
            if (horizontalTextPosition === Ui.HorizontalPosition.CENTER) {
                self.element.classList.add('p-image-paragraph-column');
                if (verticalTextPosition === Ui.VerticalPosition.TOP) {
                    if (image)
                        self.element.insertBefore(paragraph, image);
                } else if (verticalTextPosition === Ui.VerticalPosition.BOTTOM) {
                    if (image)
                        self.element.insertBefore(image, paragraph);
                } else {
                    // TODO: Think about .CENTER .CENTER
                }
            } else {
                self.element.classList.add('p-image-paragraph-row');
                if (horizontalTextPosition === Ui.HorizontalPosition.LEFT) {
                    if (image)
                        self.element.insertBefore(paragraph, image);
                } else if (horizontalTextPosition === Ui.HorizontalPosition.RIGHT) {
                    if (image)
                        self.element.insertBefore(image, paragraph);
                } // else // value of 'horizontalTextPosition' is unknown
                if (verticalTextPosition === Ui.VerticalPosition.TOP) {
                    self.element.classList.add('p-image-paragraph-row-top');
                } else if (verticalTextPosition === Ui.VerticalPosition.BOTTOM) {
                    self.element.classList.add('p-image-paragraph-row-bottom');
                }
            }
        }

        applyPosition();

        Object.defineProperty(this, "icon", {
            get: function () {
                return image;
            },
            set: function (aValue) {
                if (image !== aValue) {
                    if (image) {
                        image.classList.remove('p-image');
                        self.element.removeChild(image);
                    }
                    image = aValue;
                    if (image) {
                        self.element.appendChild(image);
                        image.classList.add('p-image');
                        applyPosition();
                    }
                }
            }
        });
        Object.defineProperty(this, "text", {
            get: function () {
                return text;
            },
            set: function (aValue) {
                text = aValue;
                paragraph.innerText = text;
            }
        });
        Object.defineProperty(this, "iconTextGap", {
            get: function () {
                return iconTextGap;
            },
            set: function (aValue) {
                iconTextGap = aValue;
            }
        });
        /**
         * Vertical position of the text relative to the icon.
         */
        Object.defineProperty(this, "horizontalTextPosition", {
            get: function () {
                return horizontalTextPosition;
            },
            set: function (aValue) {
                if (horizontalTextPosition !== aValue) {
                    horizontalTextPosition = aValue;
                    applyPosition();
                }
            }
        });

        /**
         * Vertical position of the text relative to the icon.
         */
        Object.defineProperty(this, "verticalTextPosition", {
            get: function () {
                return verticalTextPosition;
            },
            set: function (aValue) {
                if (verticalTextPosition !== aValue) {
                    verticalTextPosition = aValue;
                    applyPosition();
                }
            }
        });
    }
    extend(ImageParagraph, Widget);
    return ImageParagraph;
});